File: GUI/Materials/Output/Construction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace GUI.Materials.Output
{
    class Construction : IEquatable<Construction>
    {
        readonly private IList<string> layerNames;

        public Construction(IList<MaterialLayer> layers)
        {
            layerNames = new List<string>(layers.Select(layer => layer.Name));
        }

        public string Name
        {
            get
            {
                if (layerNames.Count == 1) { return layerNames[0]; }
                else { return String.Format("Unnamed composite (id {0})", this.GetHashCode()); }
            }
        }

        public IList<string> LayerNames { get { return layerNames; } }

        public override bool Equals(object obj)
        {
            return Equals(obj as Construction);
        }

        public bool Equals(Construction other)
        {
            if (Object.ReferenceEquals(other, null)) { return false; }
            if (Object.ReferenceEquals(this, other)) { return true; }
            if (this.LayerNames.Count != other.LayerNames.Count) { return false; }
            for (int i = 0; i < layerNames.Count; ++i)
            {
                if (this.LayerNames[i] != other.LayerNames[i]) { return false; }
            }
            return true;
        }

        public static bool operator ==(Construction a, Construction b)
        {
            if (Object.ReferenceEquals(a, null))
            {
                return Object.ReferenceEquals(b, null);
            }
            return a.Equals(b);
        }

        public static bool operator !=(Construction a, Construction b)
        {
            return !(a == b);
        }

        public override int GetHashCode()
        {
            // i use '!' as a separator because it can't legally appear in E+ construction names
            return String.Join("!", LayerNames).GetHashCode();
        }
    }
}

File: ConstructionManager/Construction.fs
namespace ConstructionManager

open System
open System.Collections.Generic

type Construction (layers:OutputLayer array) =
    let layerNames = layers |> Seq.map (fun layer -> layer.Name) |> Seq.toArray
    let identifier = String.concat "!" layerNames // "!" is the separator because it can't appear in E+ names

    member this.Name =
        if layerNames.Length = 1 then layerNames.[0]
        else sprintf "Unnamed composite (id %i)" (this.GetHashCode())

    member this.LayerNames = List<string>(layerNames)

    member private this.Identifier = identifier

    override this.Equals(obj:Object) =
        match obj with
        | :? Construction as c -> this.Equals(c)
        | _ -> false

    override this.GetHashCode () = this.Identifier.GetHashCode()

    interface IEquatable<Construction> with
        member this.Equals(other) =
            if (Object.ReferenceEquals(other, null)) then false
            else if (Object.ReferenceEquals(this, other)) then true
            else this.Identifier = other.Identifier

    interface IComparable<Construction> with
        member this.CompareTo(other) = this.Identifier.CompareTo(other.Identifier)

    interface IComparable with
        member this.CompareTo(obj) =
            match obj with
            | null -> 1
            | :? Construction as c -> (this :> IComparable<Construction>).CompareTo(c)
            | _ -> raise (ArgumentException())

